documentation/PreemptiveProvableAssertions.md

```
@@ -0,0 +1,446 @@
+
+This idea was [introduced by Nethermind](https://ethresear.ch/t/same-slot-l1-l2-message-passing/21186) and as explained in that post, it can be combined with their fast-withdrawal mechanism to perform a same-slot round-trip operation. Here I will just focus on the assert-and-prove structure of the L1-to-L2 message.
+
+As noted, the anchor block mechanism requires the Inbox contract to query the block hash of the relevant L1 block, which implies it does not support reacting to transactions included in the current L1 block. However, an L2 sequencer that can predict that a particular L1 transaction will be included in the publication block (orange in this example) can assert that claim immediately in the L2.
+
+<p align="center"><img src="./provable_assertion_images.5.png"/></p>
+
+For a same-slot deposit, the procedure would be:
+
+- a user signs a transaction that deposits to an L1 bridge contract.
+- the L2 sequencer believes this transaction will be included before their own publication transaction and it will succeed.
+    - typically this implies the L2 sequencer is also the L1 sequencer (i.e. it is a based rollup) but it could also be achieved with L1 preconfirmations.
```

markdownlint-cli2 (0.17.2)

Unordered list indentation
Expected: 2; Actual: 4

(MD007, ul-indent)